notebooks/Cross Validation Primer.ipynb

```diff
@@ -0,0 +1,432 @@
+    "\n",
+    "plt.ylim([0,6])\n",
+    "lw =2\n",
+    "\n",
+    "models = [\n",
+    "    make_pipeline(PolynomialFeatures(2), Ridge()), # Overly Simple\n",
+    "    make_pipeline(PolynomialFeatures(4), Ridge()), # Just Right\n",
+    "    DecisionTreeRegressor(max_depth=80)  # Overly Complex\n",
+    "]\n",
+    "\n",
+    "labels = ['overly simple', 'just right', 'overly complex']\n",
+    "#for count, degree in enumerate([2, 5, 12]):\n",
+    "for count, model in enumerate(models):\n",
+    "    #fit to training data\n",
+    "    model.fit(X_train, y_train)\n",
+    "    y_train_fit = model.predict(X_train)\n",
+    "    nscore = skl.metrics.r2_score(y_train, y_train_fit)\n",
+    "    #apply to validation data\n",
+    "    y_predicted = model.predict(X_val)\n",
+    "    #get R^2 score\n",
+    "    score = skl.metrics.r2_score(y_val, y_predicted)\n",
+    "    #apply again, just to show plot\n",
+    "    y_plot = model.predict(X_plot)\n",
+    "    plt.plot(x_plot, y_plot, color=colors[count], linewidth=lw, alpha=0.5,\n",
+    "             label=labels[count] + \" with CV score %0.02f \\n   naive score %0.02f\" %(score, nscore))\n",
+    "\n",
+    "plt.legend(loc='lower right')\n",
+    "\n",
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Using K-Fold Cross Validation\n",
+    "\n",
+    "The example above showed why we need a good training set and a validation set to cross validate the model. \n",
+    "If we only looked at the 'naive score', we would think the best fit is the overly-complex model because \n",
+    "it has a \"perfect score\" of 1.0.\n",
+    "\n",
+    "When we only have a few data points, we can have a good, representative training set, a good train/validate split, a good CV score and still pick the wrong model!\n",
+    "\n",
+    "Leaving a specific point in the validation set biases the cross validation score because that single, \n",
+    "specific data point is given too much importance / statistical weight. To avoid this, a method called\n",
+    "**k-fold cross validation** is employed. K-fold cross validation picks _k_ training and validation sets \n",
+    "from your data, ensuring that each point takes a turn in a training set and validation set. The average\n",
+    "of the k-fold cross validation scores are taken to indicate which model is the best.\n",
+    "\n",
+    "In the plots below, take note of the effect on CV score of having a specific data point in the validation set.\n",
+    "\n",
+    "It should be noted that there are multiple methods that achieve the same results as k-fold cross \n",
+    "validation. K-fold is one of the simplest to employ in scikit learn."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "scrolled": false
+   },
+   "outputs": [],
+   "source": [
+    "#this the same code as above\n",
```

Comment:
Abstracting the similarities into a function or a collection of functions would really help here! Don't worry about adding very bespoke flags and arguments -- these functions would be solely to make this notebook shorter and easier to read, they don't have to generalize to other contexts.